File: ios/Sources/IronSourcePlugin/IronSource.swift
import Foundation
import UIKit
import IronSource

private protocol IronSourceRewardListener: AnyObject {
    func onRewardEarned(_ payload: [String: Any])
}

private protocol IronSourceBannerListener: AnyObject {
    func onBannerEvent(_ eventName: String, payload: [String: Any])
}

private protocol IronSourceInterstitialListener: AnyObject {
    func onInterstitialEvent(_ eventName: String, payload: [String: Any])
}

private typealias IronSourceSDK = IronSource.IronSource
private typealias IronSourceInitializationDelegate = IronSource.ISInitializationDelegate
private typealias IronSourceRewardedVideoDelegate = IronSource.LevelPlayRewardedVideoDelegate
private typealias IronSourcePlacementInfo = IronSource.ISPlacementInfo
private typealias IronSourceAdInfo = IronSource.ISAdInfo
private typealias IronSourceBannerAdView = IronSource.LPMBannerAdView
private typealias IronSourceBannerAdViewConfigBuilder = IronSource.LPMBannerAdViewConfigBuilder
private typealias IronSourceBannerAdViewDelegate = IronSource.LPMBannerAdViewDelegate
private typealias IronSourceAdSize = IronSource.LPMAdSize
private typealias IronSourceInterstitialAd = IronSource.LPMInterstitialAd
private typealias IronSourceInterstitialAdDelegate = IronSource.LPMInterstitialAdDelegate
private typealias IronSourceInterstitialAdInfo = IronSource.LPMAdInfo

private final class InterstitialDelegateProxy: NSObject, IronSourceInterstitialAdDelegate {
    var onLoaded: ((IronSourceInterstitialAdInfo) -> Void)?
    var onLoadFailed: ((String, NSError) -> Void)?
    var onDisplayed: ((IronSourceInterstitialAdInfo) -> Void)?
    var onDisplayFailed: ((IronSourceInterstitialAdInfo?, NSError) -> Void)?
    var onClicked: ((IronSourceInterstitialAdInfo) -> Void)?
    var onClosed: ((IronSourceInterstitialAdInfo) -> Void)?

    func didLoadAd(withAdInfo adInfo: IronSourceInterstitialAdInfo!) {
        guard let adInfo else {
            return
        }

        onLoaded?(adInfo)
    }

    func didFailToLoadAd(withAdUnitId adUnitId: String!, error: Error!) {
        let nsError = (error as NSError?) ?? NSError(domain: "IronSource", code: -1)
        onLoadFailed?(adUnitId ?? "", nsError)
    }

    func didDisplayAd(withAdInfo adInfo: IronSourceInterstitialAdInfo!) {
        guard let adInfo else {
            return
        }

        onDisplayed?(adInfo)
    }

    func didFailToDisplayAd(withAdInfo adInfo: IronSourceInterstitialAdInfo!, error: Error!) {
        let nsError = (error as NSError?) ?? NSError(domain: "IronSource", code: -1)
        onDisplayFailed?(adInfo, nsError)
    }

    func didClickAd(withAdInfo adInfo: IronSourceInterstitialAdInfo!) {
        guard let adInfo else {
            return
        }

        onClicked?(adInfo)
    }

    func didCloseAd(withAdInfo adInfo: IronSourceInterstitialAdInfo!) {
        guard let adInfo else {
            return
        }

        onClosed?(adInfo)
    }
}

@objc public class IronSource: NSObject, IronSourceInitializationDelegate, IronSourceRewardedVideoDelegate, IronSourceBannerAdViewDelegate {
    private weak var rewardListener: IronSourceRewardListener?
    private weak var bannerListener: IronSourceBannerListener?
    private weak var interstitialListener: IronSourceInterstitialListener?
    private var initializeCompletion: (() -> Void)?
    private var bannerAdView: IronSourceBannerAdView?
    private var bannerAdUnitId: String?
    private var interstitialAd: IronSourceInterstitialAd?
    private var interstitialAdUnitId: String?
    private var interstitialPlacementName: String?
    private let interstitialDelegateProxy = InterstitialDelegateProxy()

    private(set) var initialized = false
    private(set) var rewardedVideoAvailable = false
    private(set) var interstitialAvailable = false

    init(
        rewardListener: IronSourceRewardListener,
        bannerListener: IronSourceBannerListener,
        interstitialListener: IronSourceInterstitialListener
    ) {
        self.rewardListener = rewardListener
        self.bannerListener = bannerListener
        self.interstitialListener = interstitialListener
        super.init()

        interstitialDelegateProxy.onLoaded = { [weak self] adInfo in
            self?.handleInterstitialLoaded(adInfo)
        }
        interstitialDelegateProxy.onLoadFailed = { [weak self] adUnitId, error in
            self?.handleInterstitialLoadFailed(adUnitId: adUnitId, error: error)
        }
        interstitialDelegateProxy.onDisplayed = { [weak self] adInfo in
            self?.handleInterstitialDisplayed(adInfo)
        }
        interstitialDelegateProxy.onDisplayFailed = { [weak self] adInfo, error in
            self?.handleInterstitialDisplayFailed(adInfo: adInfo, error: error)
        }
        interstitialDelegateProxy.onClicked = { [weak self] adInfo in
            self?.handleInterstitialClicked(adInfo)
        }
        interstitialDelegateProxy.onClosed = { [weak self] adInfo in
            self?.handleInterstitialClosed(adInfo)
        }
    }

    func initialize(appKey: String, completion: @escaping () -> Void) {
        initializeCompletion = completion
        initialized = false
        rewardedVideoAvailable = false
        interstitialAvailable = false
        interstitialAd = nil
        interstitialAdUnitId = nil
        interstitialPlacementName = nil

        DispatchQueue.main.async { [weak self] in
            guard let self else {
                return
            }

            IronSourceSDK.setLevelPlayRewardedVideoDelegate(self)
            IronSourceSDK.init(withAppKey: appKey, adUnits: [IS_REWARDED_VIDEO, IS_BANNER], delegate: self)
        }
    }

    func setMetaData(key: String, value: String) {
        DispatchQueue.main.async {
            LevelPlay.setMetaDataWithKey(key, value: value)
        }
    }

    func setMetaData(key: String, values: [String]) {
        DispatchQueue.main.async {
            LevelPlay.setMetaDataWithKey(key, values: NSMutableArray(array: values))
        }
    }

    func launchTestSuite(from viewController: UIViewController) {
        DispatchQueue.main.async {
            LevelPlay.launchTestSuite(viewController)
        }
    }

    func loadRewardedVideo() {
        DispatchQueue.main.async {
            IronSourceSDK.loadRewardedVideo()
        }
    }

    func isRewardedVideoReady() -> Bool {
        rewardedVideoAvailable && IronSourceSDK.hasRewardedVideo()
    }

    func showRewardedVideo(from viewController: UIViewController) {
        DispatchQueue.main.async {
            IronSourceSDK.showRewardedVideo(with: viewController)
        }
    }

    func loadInterstitial(adUnitId: String, placementName: String) {
        destroyInterstitial()

        let interstitialAd = IronSourceInterstitialAd(adUnitId: adUnitId)
        interstitialAd.setDelegate(interstitialDelegateProxy)
        self.interstitialAd = interstitialAd
        interstitialAdUnitId = adUnitId
        interstitialPlacementName = placementName
        interstitialAvailable = interstitialAd.isAdReady()

        DispatchQueue.main.async {
            interstitialAd.loadAd()
        }
    }

    func isInterstitialReady() -> Bool {
        interstitialAvailable && (interstitialAd?.isAdReady() ?? false)
    }

    func showInterstitial(from viewController: UIViewController) {
        guard let interstitialAd else {
            return
        }

        let placementName = interstitialPlacementName?.isEmpty == false ? interstitialPlacementName : nil
        DispatchQueue.main.async {
            interstitialAd.showAd(viewController: viewController, placementName: placementName)
        }
    }

    func destroyInterstitial() {
        interstitialAd?.setDelegate(nil)
        interstitialAd = nil
        interstitialAvailable = false
        interstitialAdUnitId = nil
        interstitialPlacementName = nil
    }

    func loadBanner(
        adUnitId: String,
        placementName: String,
        in viewController: UIViewController,
        webView: UIView?,
        x: Double,
        y: Double,
        width: Double,
        height: Double,
        scale: Double
    ) {
        destroyBanner()

        let adSize = IronSourceAdSize.createAdaptiveAdSize()
        let configBuilder = IronSourceBannerAdViewConfigBuilder()
        configBuilder.setWithAdSize(adSize)
        if !placementName.isEmpty {
            configBuilder.setWithPlacementName(placementName)
        }

        let bannerView = IronSourceBannerAdView(adUnitId: adUnitId, config: configBuilder.build())
        bannerView.delegate = self
        bannerView.isHidden = true
        bannerView.translatesAutoresizingMaskIntoConstraints = true

        viewController.view.addSubview(bannerView)
        bannerAdView = bannerView
        bannerAdUnitId = adUnitId

        updateBannerLayout(webView: webView, x: x, y: y, width: width, height: height, scale: scale)
        bannerView.loadAd(with: viewController)
    }

    func updateBannerLayout(webView: UIView?, x: Double, y: Double, width: Double, height: Double, scale: Double) {
        guard let bannerAdView else {
            return
        }

        let safeScale = scale > 0 ? scale : 1
        let webViewOrigin = webView?.frame.origin ?? .zero
        let frame = CGRect(
            x: webViewOrigin.x + CGFloat(x * safeScale),
            y: webViewOrigin.y + CGFloat(y * safeScale),
            width: max(CGFloat(width * safeScale), 1),
            height: max(CGFloat(height * safeScale), 1)
        )

        bannerAdView.frame = frame
    }

    func showBanner() {
        bannerAdView?.isHidden = false
    }

    func hideBanner() {
        bannerAdView?.isHidden = true
    }

    func destroyBanner() {
        guard let bannerAdView else {
            return
        }

        bannerAdView.removeFromSuperview()
        bannerAdView.destroy()
        self.bannerAdView = nil
        bannerAdUnitId = nil
    }

    private func buildBannerPayload(adInfo: IronSourceAdInfo?) -> [String: Any] {
        var payload: [String: Any] = [:]

        if let bannerAdUnitId {
            payload["adUnitId"] = bannerAdUnitId
        }

        if let frame = bannerAdView?.frame {
            payload["width"] = Int(frame.width)
            payload["height"] = Int(frame.height)
        }

        if let adInfo {
            payload["adUnitId"] = adInfo.adUnit
            payload["adNetwork"] = adInfo.adNetwork
            payload["instanceId"] = adInfo.instanceId
            payload["instanceName"] = adInfo.instanceName
        }

        return payload
    }

    private func buildInterstitialPayload(adInfo: IronSourceInterstitialAdInfo?) -> [String: Any] {
        var payload: [String: Any] = [:]

        if let interstitialAdUnitId {
            payload["adUnitId"] = interstitialAdUnitId
        }

        if let interstitialPlacementName, !interstitialPlacementName.isEmpty {
            payload["placement"] = interstitialPlacementName
        }

        if let adInfo {
            payload["placement"] = adInfo.placementName ?? payload["placement"]
            payload["adUnit"] = adInfo.adUnitId
            payload["adNetwork"] = adInfo.adNetwork
            payload["instanceId"] = adInfo.instanceId
            payload["instanceName"] = adInfo.instanceName
        }

        return payload
    }

    private func buildInterstitialErrorPayload(adInfo: IronSourceInterstitialAdInfo?, error: NSError) -> [String: Any] {
        var payload = buildInterstitialPayload(adInfo: adInfo)
        payload["errorCode"] = error.code
        payload["errorMessage"] = error.localizedDescription
        return payload
    }

    private func handleInterstitialLoaded(_ adInfo: IronSourceInterstitialAdInfo) {
        interstitialAvailable = true
        interstitialListener?.onInterstitialEvent("interstitialLoaded", payload: buildInterstitialPayload(adInfo: adInfo))
    }

    private func handleInterstitialLoadFailed(adUnitId: String, error: NSError) {
        interstitialAvailable = false
        if !adUnitId.isEmpty {
            interstitialAdUnitId = adUnitId
        }
        interstitialListener?.onInterstitialEvent(
            "interstitialLoadFailed",
            payload: buildInterstitialErrorPayload(adInfo: nil, error: error)
        )
    }

    private func handleInterstitialDisplayed(_ adInfo: IronSourceInterstitialAdInfo) {
        interstitialAvailable = false
        interstitialListener?.onInterstitialEvent("interstitialDisplayed", payload: buildInterstitialPayload(adInfo: adInfo))
    }

    private func handleInterstitialDisplayFailed(adInfo: IronSourceInterstitialAdInfo?, error: NSError) {
        interstitialAvailable = interstitialAd?.isAdReady() ?? false
        interstitialListener?.onInterstitialEvent(
            "interstitialDisplayFailed",
            payload: buildInterstitialErrorPayload(adInfo: adInfo, error: error)
        )
    }

    private func handleInterstitialClicked(_ adInfo: IronSourceInterstitialAdInfo) {
        interstitialListener?.onInterstitialEvent("interstitialClicked", payload: buildInterstitialPayload(adInfo: adInfo))
    }

    private func handleInterstitialClosed(_ adInfo: IronSourceInterstitialAdInfo) {
        interstitialAvailable = interstitialAd?.isAdReady() ?? false
        interstitialListener?.onInterstitialEvent("interstitialClosed", payload: buildInterstitialPayload(adInfo: adInfo))
    }

    public func initializationDidComplete() {
        initialized = true

        let completion = initializeCompletion
        initializeCompletion = nil

        DispatchQueue.main.async {
            completion?()
        }
    }

    public func hasAvailableAdWithAdInfo(_ adInfo: IronSourceAdInfo!) {
        rewardedVideoAvailable = true
    }

    public func hasNoAvailableAd() {
        rewardedVideoAvailable = false
    }

    public func didReceiveRewardForPlacement(_ placementInfo: IronSourcePlacementInfo!, withAdInfo adInfo: IronSourceAdInfo!) {
        var payload: [String: Any] = [
            "placement": placementInfo?.placementName ?? "",
            "rewardName": placementInfo?.rewardName ?? "",
            "amount": placementInfo?.rewardAmount?.intValue ?? 0,
        ]

        if let adInfo {
            payload["adUnit"] = adInfo.adUnit
            payload["adNetwork"] = adInfo.adNetwork
            payload["instanceId"] = adInfo.instanceId
            payload["instanceName"] = adInfo.instanceName
        }

        DispatchQueue.main.async { [weak self] in
            self?.rewardListener?.onRewardEarned(payload)
        }
    }

    public func didFailToShowWithError(_ error: Error!, andAdInfo adInfo: IronSourceAdInfo!) {
        rewardedVideoAvailable = IronSourceSDK.hasRewardedVideo()
    }

    public func didOpenWithAdInfo(_ adInfo: IronSourceAdInfo!) {
        rewardedVideoAvailable = false
    }

    public func didClick(_ placementInfo: IronSourcePlacementInfo!, withAdInfo adInfo: IronSourceAdInfo!) {
    }

    public func didCloseWithAdInfo(_ adInfo: IronSourceAdInfo!) {
        rewardedVideoAvailable = IronSourceSDK.hasRewardedVideo()
    }

    public func didLoadAd(withAdInfo adInfo: IronSourceAdInfo!) {
        bannerListener?.onBannerEvent("bannerLoaded", payload: buildBannerPayload(adInfo: adInfo))
    }

    public func didFailToLoadAd(withAdUnitId adUnitId: String!, error: Error!) {
        var payload = buildBannerPayload(adInfo: nil)
        payload["adUnitId"] = adUnitId ?? bannerAdUnitId ?? ""
        payload["errorMessage"] = error.localizedDescription
        bannerListener?.onBannerEvent("bannerLoadFailed", payload: payload)
    }

    public func didClickAd(withAdInfo adInfo: IronSourceAdInfo!) {
        bannerListener?.onBannerEvent("bannerClicked", payload: buildBannerPayload(adInfo: adInfo))
    }

    public func didDisplayAd(withAdInfo adInfo: IronSourceAdInfo!) {
        bannerListener?.onBannerEvent("bannerDisplayed", payload: buildBannerPayload(adInfo: adInfo))
    }

    public func didFailToDisplayAd(withAdInfo adInfo: IronSourceAdInfo!, error: Error!) {
        var payload = buildBannerPayload(adInfo: adInfo)
        payload["errorMessage"] = error.localizedDescription
        bannerListener?.onBannerEvent("bannerLoadFailed", payload: payload)
    }
}

File: android/src/main/java/com/juliocodex/capacitor/ironsource/IronSource.java
package com.juliocodex.capacitor.ironsource;

import android.app.Activity;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;
import android.widget.FrameLayout;

import com.getcapacitor.JSObject;
import com.unity3d.mediation.LevelPlay;
import com.unity3d.mediation.LevelPlayAdInfo;
import com.unity3d.mediation.LevelPlayAdError;
import com.unity3d.mediation.LevelPlayAdSize;
import com.unity3d.mediation.LevelPlayConfiguration;
import com.unity3d.mediation.LevelPlayInitError;
import com.unity3d.mediation.LevelPlayInitListener;
import com.unity3d.mediation.LevelPlayInitRequest;
import com.unity3d.mediation.banner.LevelPlayBannerAdView;
import com.unity3d.mediation.banner.LevelPlayBannerAdViewListener;
import com.unity3d.mediation.interstitial.LevelPlayInterstitialAd;
import com.unity3d.mediation.interstitial.LevelPlayInterstitialAdListener;
import com.unity3d.mediation.rewarded.LevelPlayReward;
import com.unity3d.mediation.rewarded.LevelPlayRewardedAd;
import com.unity3d.mediation.rewarded.LevelPlayRewardedAdListener;

import java.util.List;

public class IronSource implements LevelPlayRewardedAdListener {

    public interface InitializationErrorListener {
        void onInitializationFailed(String errorMessage);
    }

    public interface RewardEventListener {
        void onRewardEarned(JSObject payload);
    }

    public interface RewardedStatusListener {
        void onRewardedEvent(String eventName, JSObject payload);
    }

    public interface BannerEventListener {
        void onBannerEvent(String eventName, JSObject payload);
    }

    public interface InterstitialEventListener {
        void onInterstitialEvent(String eventName, JSObject payload);
    }

    private final RewardEventListener rewardEventListener;
    private final RewardedStatusListener rewardedStatusListener;
    private final BannerEventListener bannerEventListener;
    private final InterstitialEventListener interstitialEventListener;
    private volatile boolean initialized;
    private volatile boolean rewardedVideoAvailable;
    private volatile boolean interstitialAvailable;
    private LevelPlayRewardedAd rewardedAd;
    private String rewardedAdUnitId;
    private LevelPlayInterstitialAd interstitialAd;
    private String interstitialAdUnitId;
    private String interstitialPlacementName;
    private LevelPlayBannerAdView bannerAdView;
    private ViewGroup bannerContainer;
    private String bannerAdUnitId;
    private String bannerPlacementName;
    private final LevelPlayBannerAdViewListener bannerAdListener = new LevelPlayBannerAdViewListener() {
        @Override
        public void onAdLoaded(LevelPlayAdInfo adInfo) {
            bannerEventListener.onBannerEvent("bannerLoaded", buildBannerPayload(adInfo));
        }

        @Override
        public void onAdLoadFailed(LevelPlayAdError error) {
            bannerEventListener.onBannerEvent("bannerLoadFailed", buildBannerErrorPayload(null, error));
        }

        @Override
        public void onAdDisplayed(LevelPlayAdInfo adInfo) {
            bannerEventListener.onBannerEvent("bannerDisplayed", buildBannerPayload(adInfo));
        }

        @Override
        public void onAdDisplayFailed(LevelPlayAdInfo adInfo, LevelPlayAdError error) {
            bannerEventListener.onBannerEvent("bannerLoadFailed", buildBannerErrorPayload(adInfo, error));
        }

        @Override
        public void onAdClicked(LevelPlayAdInfo adInfo) {
            bannerEventListener.onBannerEvent("bannerClicked", buildBannerPayload(adInfo));
        }
    };
    private final LevelPlayInterstitialAdListener interstitialAdListener = new LevelPlayInterstitialAdListener() {
        @Override
        public void onAdLoaded(LevelPlayAdInfo adInfo) {
            interstitialAvailable = true;
            interstitialEventListener.onInterstitialEvent("interstitialLoaded", buildInterstitialPayload(adInfo));
        }

        @Override
        public void onAdLoadFailed(LevelPlayAdError error) {
            interstitialAvailable = false;
            interstitialEventListener.onInterstitialEvent("interstitialLoadFailed", buildInterstitialErrorPayload(error, null));
        }

        @Override
        public void onAdDisplayed(LevelPlayAdInfo adInfo) {
            interstitialAvailable = false;
            interstitialEventListener.onInterstitialEvent("interstitialDisplayed", buildInterstitialPayload(adInfo));
        }

        @Override
        public void onAdDisplayFailed(LevelPlayAdError error, LevelPlayAdInfo adInfo) {
            interstitialAvailable = interstitialAd != null && interstitialAd.isAdReady();
            interstitialEventListener.onInterstitialEvent("interstitialDisplayFailed", buildInterstitialErrorPayload(error, adInfo));
        }

        @Override
        public void onAdClicked(LevelPlayAdInfo adInfo) {
            interstitialEventListener.onInterstitialEvent("interstitialClicked", buildInterstitialPayload(adInfo));
        }

        @Override
        public void onAdClosed(LevelPlayAdInfo adInfo) {
            interstitialAvailable = interstitialAd != null && interstitialAd.isAdReady();
            interstitialEventListener.onInterstitialEvent("interstitialClosed", buildInterstitialPayload(adInfo));
        }

        @Override
        public void onAdInfoChanged(LevelPlayAdInfo adInfo) {
        }
    };

    public IronSource(
        RewardEventListener rewardEventListener,
        RewardedStatusListener rewardedStatusListener,
        BannerEventListener bannerEventListener,
        InterstitialEventListener interstitialEventListener
    ) {
        this.rewardEventListener = rewardEventListener;
        this.rewardedStatusListener = rewardedStatusListener;
        this.bannerEventListener = bannerEventListener;
        this.interstitialEventListener = interstitialEventListener;
    }

    public void initialize(
        Activity activity,
        String appKey,
        String userId,
        Runnable onInitializationComplete,
        InitializationErrorListener onInitializationFailed
    ) {
        initialized = false;
        rewardedVideoAvailable = false;
        rewardedAd = null;
        rewardedAdUnitId = null;
        interstitialAvailable = false;
        interstitialAd = null;
        interstitialAdUnitId = null;
        interstitialPlacementName = null;

        LevelPlayInitRequest.Builder requestBuilder = new LevelPlayInitRequest.Builder(appKey);

        if (userId != null && !userId.isEmpty()) {
            requestBuilder.withUserId(userId);
        }

        LevelPlay.init(activity, requestBuilder.build(), new LevelPlayInitListener() {
            @Override
            public void onInitSuccess(LevelPlayConfiguration configuration) {
                initialized = true;
                if (onInitializationComplete != null) {
                    onInitializationComplete.run();
                }
            }

            @Override
            public void onInitFailed(LevelPlayInitError error) {
                initialized = false;
                if (onInitializationFailed != null) {
                    String errorMessage = error != null ? error.getErrorMessage() : "LevelPlay initialization failed.";
                    onInitializationFailed.onInitializationFailed(errorMessage);
                }
            }
        });
    }

    public boolean isInitialized() {
        return initialized;
    }

    public void setMetaData(String key, String value) {
        LevelPlay.setMetaData(key, value);
    }

    public void setMetaData(String key, List<String> values) {
        LevelPlay.setMetaData(key, values);
    }

    public void launchTestSuite(Activity activity) {
        LevelPlay.launchTestSuite(activity);
    }

    public void loadRewardedVideo(String adUnitId) {
        if (rewardedAd == null || !adUnitId.equals(rewardedAdUnitId)) {
            rewardedAd = new LevelPlayRewardedAd(adUnitId);
            rewardedAd.setListener(this);
            rewardedAdUnitId = adUnitId;
        }

        rewardedVideoAvailable = rewardedAd.isAdReady();
        rewardedAd.loadAd();
    }

    public boolean isRewardedVideoReady() {
        return rewardedAd != null && rewardedAd.isAdReady();
    }

    public void showRewardedVideo(Activity activity) {
        if (rewardedAd == null) {
            return;
        }

        rewardedAd.showAd(activity);
    }

    public void loadInterstitial(String adUnitId, String placementName) {
        if (interstitialAd == null || !adUnitId.equals(interstitialAdUnitId)) {
            interstitialAd = new LevelPlayInterstitialAd(adUnitId);
            interstitialAd.setListener(interstitialAdListener);
            interstitialAdUnitId = adUnitId;
        }

        interstitialPlacementName = placementName;
        interstitialAvailable = interstitialAd.isAdReady();
        interstitialAd.loadAd();
    }

    public boolean isInterstitialReady() {
        return interstitialAd != null && interstitialAd.isAdReady();
    }

    public void showInterstitial(Activity activity) {
        if (interstitialAd == null) {
            return;
        }

        interstitialAd.showAd(activity);
    }

    public void destroyInterstitial() {
        if (interstitialAd == null) {
            return;
        }

        interstitialAd.setListener(null);
        interstitialAd = null;
        interstitialAvailable = false;
        interstitialAdUnitId = null;
        interstitialPlacementName = null;
    }

    public void loadBanner(Activity activity, View webView, String adUnitId, String placementName, double x, double y, double width, double height, double scale) {
        destroyBanner();

        bannerAdUnitId = adUnitId;
        bannerPlacementName = placementName;

        LevelPlayBannerAdView.Config.Builder configBuilder = new LevelPlayBannerAdView.Config.Builder()
            .setAdSize(LevelPlayAdSize.createAdaptiveAdSize(activity));
        if (!placementName.isEmpty()) {
            configBuilder.setPlacementName(placementName);
        }

        bannerAdView = new LevelPlayBannerAdView(activity, adUnitId, configBuilder.build());
        bannerAdView.setBannerListener(bannerAdListener);
        bannerAdView.setVisibility(View.GONE);

        bannerContainer = activity.findViewById(android.R.id.content);
        attachBannerView();

        updateBannerLayout(webView, x, y, width, height, scale);
        bannerAdView.loadAd();
    }

    public void updateBannerLayout(View webView, double x, double y, double width, double height, double scale) {
        if (bannerAdView == null || webView == null) {
            return;
        }

        float safeScale = scale > 0 ? (float) scale : 1f;
        int left = Math.round((float) x * safeScale) + webView.getLeft();
        int top = Math.round((float) y * safeScale) + webView.getTop();
        int layoutWidth = Math.max(Math.round((float) width * safeScale), 1);
        int layoutHeight = Math.max(Math.round((float) height * safeScale), 1);

        FrameLayout.LayoutParams layoutParams = new FrameLayout.LayoutParams(layoutWidth, layoutHeight);
        layoutParams.leftMargin = left;
        layoutParams.topMargin = top;
        bannerAdView.setLayoutParams(layoutParams);
        bannerAdView.requestLayout();
    }

    public void showBanner(Activity activity) {
        if (bannerAdView == null) {
            return;
        }

        if (bannerContainer == null && activity != null) {
            bannerContainer = activity.findViewById(android.R.id.content);
        }

        attachBannerView();
        bannerAdView.setVisibility(View.VISIBLE);
        bannerAdView.bringToFront();
        bannerAdView.requestLayout();
        bannerAdView.invalidate();
    }

    public void hideBanner() {
        if (bannerAdView == null) {
            return;
        }

        bannerAdView.setVisibility(View.GONE);
        detachBannerView();
    }

    public void destroyBanner() {
        if (bannerAdView == null) {
            return;
        }

        detachBannerView();

        bannerAdView.setBannerListener(null);
        bannerAdView.destroy();
        bannerAdView = null;
        bannerContainer = null;
        bannerAdUnitId = null;
        bannerPlacementName = null;
    }

    private void attachBannerView() {
        if (bannerAdView == null || bannerContainer == null) {
            return;
        }

        ViewParent parent = bannerAdView.getParent();
        if (parent == bannerContainer) {
            return;
        }

        if (parent instanceof ViewGroup) {
            ((ViewGroup) parent).removeView(bannerAdView);
        }

        bannerContainer.addView(bannerAdView);
    }

    private void detachBannerView() {
        if (bannerAdView == null) {
            return;
        }

        ViewParent parent = bannerAdView.getParent();
        if (parent instanceof ViewGroup) {
            ((ViewGroup) parent).removeView(bannerAdView);
        }
    }

    private JSObject buildBannerPayload(LevelPlayAdInfo adInfo) {
        JSObject payload = new JSObject();

        if (bannerAdView != null) {
            payload.put("width", bannerAdView.getWidth());
            payload.put("height", bannerAdView.getHeight());
        }

        if (adInfo == null) {
            if (bannerAdUnitId != null) {
                payload.put("adUnitId", bannerAdUnitId);
            }
            if (bannerPlacementName != null && !bannerPlacementName.isEmpty()) {
                payload.put("placement", bannerPlacementName);
            }
            return payload;
        }

        if (bannerAdUnitId != null) {
            payload.put("adUnitId", bannerAdUnitId);
        }
        if (bannerPlacementName != null && !bannerPlacementName.isEmpty()) {
            payload.put("placement", bannerPlacementName);
        }
        if (adInfo.getPlacementName() != null && !adInfo.getPlacementName().isEmpty()) {
            payload.put("placement", adInfo.getPlacementName());
        }
        payload.put("adUnit", adInfo.getAdUnitId());
        payload.put("adFormat", "BANNER");
        payload.put("adNetwork", adInfo.getAdNetwork());
        payload.put("instanceId", adInfo.getInstanceId());
        payload.put("instanceName", adInfo.getInstanceName());
        return payload;
    }

    private JSObject buildBannerErrorPayload(LevelPlayAdInfo adInfo, LevelPlayAdError error) {
        JSObject payload = buildBannerPayload(adInfo);
        if (error != null) {
            payload.put("errorCode", error.getErrorCode());
            payload.put("errorMessage", error.getErrorMessage());
        }
        return payload;
    }

    private JSObject buildRewardPayload(LevelPlayReward reward, LevelPlayAdInfo adInfo) {
        JSObject payload = new JSObject();
        payload.put("rewardName", reward.getName());
        payload.put("amount", reward.getAmount());

        if (adInfo != null) {
            payload.put("placement", adInfo.getPlacementName());
            payload.put("adUnit", adInfo.getAdUnitId());
            payload.put("adNetwork", adInfo.getAdNetwork());
            payload.put("instanceId", adInfo.getInstanceId());
            payload.put("instanceName", adInfo.getInstanceName());
        }

        return payload;
    }

    private JSObject buildRewardedPayload(LevelPlayAdInfo adInfo) {
        JSObject payload = new JSObject();

        if (rewardedAdUnitId != null) {
            payload.put("adUnitId", rewardedAdUnitId);
        }

        if (adInfo == null) {
            return payload;
        }

        payload.put("placement", adInfo.getPlacementName());
        payload.put("adUnit", adInfo.getAdUnitId());
        payload.put("adNetwork", adInfo.getAdNetwork());
        payload.put("instanceId", adInfo.getInstanceId());
        payload.put("instanceName", adInfo.getInstanceName());
        return payload;
    }

    private JSObject buildRewardedErrorPayload(LevelPlayAdError error, LevelPlayAdInfo adInfo) {
        JSObject payload = buildRewardedPayload(adInfo);
        if (error != null) {
            payload.put("errorCode", error.getErrorCode());
            payload.put("errorMessage", error.getErrorMessage());
        }
        return payload;
    }

    private JSObject buildInterstitialPayload(LevelPlayAdInfo adInfo) {
        JSObject payload = new JSObject();

        if (interstitialAdUnitId != null) {
            payload.put("adUnitId", interstitialAdUnitId);
        }

        if (interstitialPlacementName != null && !interstitialPlacementName.isEmpty()) {
            payload.put("placement", interstitialPlacementName);
        }

        if (adInfo == null) {
            return payload;
        }

        payload.put("placement", adInfo.getPlacementName());
        payload.put("adUnit", adInfo.getAdUnitId());
        payload.put("adNetwork", adInfo.getAdNetwork());
        payload.put("instanceId", adInfo.getInstanceId());
        payload.put("instanceName", adInfo.getInstanceName());
        return payload;
    }

    private JSObject buildInterstitialErrorPayload(LevelPlayAdError error, LevelPlayAdInfo adInfo) {
        JSObject payload = buildInterstitialPayload(adInfo);
        if (error != null) {
            payload.put("errorCode", error.getErrorCode());
            payload.put("errorMessage", error.getErrorMessage());
        }
        return payload;
    }

    @Override
    public void onAdLoaded(LevelPlayAdInfo adInfo) {
        rewardedVideoAvailable = true;
        rewardedStatusListener.onRewardedEvent("rewardedLoaded", buildRewardedPayload(adInfo));
    }

    @Override
    public void onAdLoadFailed(LevelPlayAdError error) {
        rewardedVideoAvailable = false;
        rewardedStatusListener.onRewardedEvent("rewardedLoadFailed", buildRewardedErrorPayload(error, null));
    }

    @Override
    public void onAdDisplayed(LevelPlayAdInfo adInfo) {
        rewardedVideoAvailable = false;
    }

    @Override
    public void onAdRewarded(LevelPlayReward reward, LevelPlayAdInfo adInfo) {
        rewardEventListener.onRewardEarned(buildRewardPayload(reward, adInfo));
    }

    @Override
    public void onAdDisplayFailed(LevelPlayAdError error, LevelPlayAdInfo adInfo) {
        rewardedVideoAvailable = rewardedAd != null && rewardedAd.isAdReady();
    }

    @Override
    public void onAdClicked(LevelPlayAdInfo adInfo) {
    }

    @Override
    public void onAdClosed(LevelPlayAdInfo adInfo) {
        rewardedVideoAvailable = rewardedAd != null && rewardedAd.isAdReady();
    }

    @Override
    public void onAdInfoChanged(LevelPlayAdInfo adInfo) {
    }

}
